cpp/src/arrow/engine/substrait/serde.h

```diff
@@ -115,6 +115,22 @@ ARROW_ENGINE_EXPORT Result<compute::ExecPlan> DeserializePlan(
     const Buffer& buf, const std::shared_ptr<dataset::WriteNodeOptions>& write_options,
     const ExtensionIdRegistry* registry = NULLPTR, ExtensionSet* ext_set_out = NULLPTR);
 
+/// Factory function type for generating the write options of a node consuming the batches
+/// produced by each toplevel Substrait relation when deserializing a Substrait Plan.
+using WriteOptionsFactory = std::function<std::shared_ptr<dataset::WriteNodeOptions>()>;
+
+struct ARROW_ENGINE_EXPORT UdfDeclaration {
+  std::string name;
+  std::string code;
+  std::string summary;
+  std::string description;
+  std::vector<std::pair<std::shared_ptr<DataType>, bool>> input_types;
+  std::pair<std::shared_ptr<DataType>, bool> output_type;
```

Comment:
It's not obvious what the `bool` is for. Can you perhaps use a helper struct, e.g.:
```suggestion
  struct TypeDeclaration {
    std::shared_ptr<DataType> type;
    bool xxx_some_suitable_name;
  };
  std::vector<TypeDeclaration> input_types;
  TypeDeclaration output_type;
```